src/lib.rs

```diff
@@ -115,34 +115,45 @@ unsafe fn console_on_any(fds: &[DWORD]) -> bool {
 
 /// Returns true if there is an MSYS tty on the given handle.
 #[cfg(windows)]
-unsafe fn msys_tty_on(fd: DWORD) -> bool {
-    use std::{mem, slice};
-
+fn msys_tty_on(fd: DWORD) -> bool {
```

Comment:
I suspect this should be `unsafe`. It's taking a raw file descriptor without any guarantees about the validity of that file descriptor.

Reply:
The `fd` isn't used as a file descriptor internally, just as an argument to `GetStdHandle` which maps the handle id to the real runtime handle. I've documented why that's sound in its own `unsafe` block, which I'm fairly confident upholds `winapi`s contracts.

Most `winapi` APIs are actually very forgiving with handles, and will simply return errors when they're invalild.

Reply:
@sunfishcode What do you think about this? Particularly with respect to your new I/O safety RFC, should a routine like this be marked as `unsafe` even though there is no possibility of UB for any value of `fd`?

Reply:
This... is a subtle one. It's true there's no possibility of UB here, or even corrupted I/O. On the other hand, the overall pattern here of using a pre-existing I/O handle without being given explicit access to it is the kind of thing I/O safety wants to avoid. On the first hand though, atty's own public API does this, by taking an enum instead of a handle; it's not `unsafe`, and this is doing essentially the same thing.

So I think it's reasonable to leave this as a safe function for now. But I also think it makes sense to explore APIs where the user passes in a handle.